internal/chart_test.go

```diff
@@ -0,0 +1,306 @@
+package ledger
+
+import (
+	"encoding/json"
+
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func expectInvalidChart(t *testing.T, source string, expectedError string) {
+	var chart ChartOfAccounts
+	err := json.Unmarshal([]byte(source), &chart)
+
+	require.ErrorContains(t, err, expectedError)
+}
+
+func TestChartOfAccounts(t *testing.T) {
+	source := `{
+    "banks": {
+        "$iban": {
+            ".pattern": "[0-9]{10}",
+            "main": {
+                ".rules": {}
+            },
+            "out": {
+                ".metadata": {
+                    "key": "value"
+                },
+                ".rules": {}
+            },
+            "pending_out": {
+                ".rules": {}
+            }
+        }
+    },
+    "users": {
+        "$userID": {
+            ".self": {},
+            ".pattern": "[0-9]{5}",
+            ".rules": {},
+            "main": {
+                ".rules": {}
+            }
+        }
+    }
+}`
+	expected := ChartOfAccounts{
+		"banks": {
+			VariableSegment: &ChartVariableSegment{
+				Label:   "iban",
+				Pattern: "[0-9]{10}",
+				ChartSegment: ChartSegment{
+					FixedSegments: map[string]ChartSegment{
+						"main": {
+							Account: &ChartAccount{
+								Rules: ChartAccountRules{},
+							},
+						},
+						"out": {
+							Account: &ChartAccount{
+								Metadata: map[string]string{
+									"key": "value",
+								},
+							},
+						},
+						"pending_out": {
+							Account: &ChartAccount{},
+						},
+					},
+				},
+			},
+		},
+		"users": {
+			VariableSegment: &ChartVariableSegment{
+				Label:   "userID",
+				Pattern: "[0-9]{5}",
+				ChartSegment: ChartSegment{
+					Account: &ChartAccount{},
+					FixedSegments: map[string]ChartSegment{
+						"main": {
+							Account: &ChartAccount{},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	var chart ChartOfAccounts
+	err := json.Unmarshal([]byte(source), &chart)
+	require.NoError(t, err)
+
+	require.Equal(t, expected, chart)
+
+	value, err := json.MarshalIndent(&chart, "", "    ")
+	require.NoError(t, err)
+	require.JSONEq(t, source, string(value))
+}
+
+func TestInvalidFixedSegment(t *testing.T) {
+	src := `{
```

Comment:
Not big deal, but please use t.Parallel() when you can (in a perfect world : every time)